### PR TITLE
[GPU] Update config api 2.0

### PR DESCRIPTION
**Pull request by sshlyapn**
This patch updates GPU Plugin's:

- [x] Configuration keys
- [x] Metrics keys
- [x] Params keys

### CODE REVIEW

File: src/plugins/intel_gpu/src/plugin/plugin.cpp [hunk 1]
```diff
@@ -606,13 +610,42 @@ static float GetGOPS(cldnn::device_info info, cldnn::data_types dt) {
...
     auto device = iter != device_map.end() ? iter->second : device_map.begin()->second;
     auto device_info = device->get_info();
-
-    if (name == METRIC_KEY(SUPPORTED_METRICS)) {
+    bool is_new_api = GetCore()->isNewAPI();
+
+    if (name == ov::supported_properties) {
+        return decltype(ov::supported_properties)::value_type {
+        // Metrices
+        ov::supported_properties.name(),
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/plugin/plugin.cpp:623]: I think we need to set mutability as well for each property. Like
```
PropertyName{ov::supported_properties.name(), PropertyMutability::RO},
PropertyName{ov::available_devices.name(), PropertyMutability::RO},
etc
```

File: src/plugins/intel_gpu/src/plugin/device_config.cpp [hunk 2]
```diff
@@ -114,50 +117,72 @@ void Config::UpdateFromMap(const std::map<std::string, std::string>& configMap)
                 default:
                     IE_THROW(ParameterMismatch) << "Unsupported queue priority value: " << uVal;
             }
-        } else if (key.compare(GPUConfigParams::KEY_GPU_MODEL_PRIORITY) == 0) {
+        } else if (key.compare(GPUConfigParams::KEY_GPU_MODEL_PRIORITY) == 0 ||
+                   key.compare(ov::hint::model_priority.name()) == 0) {
+            const bool new_api = key.compare(ov::hint::model_priority.name()) == 0;
             bool found_matched_value = false;
-            if (val.find(GPUConfigParams::GPU_MODEL_PRIORITY_HIGH) != std::string::npos) {
```
[the remark this answers is not in the document]
Comment (sshlyapn) [reply src/plugins/intel_gpu/src/plugin/device_config.cpp:119]: Do I understand correctly that you are talking about replacing `DECLARE_GPU_CONFIG_KEY(MODEL_PRIORITY)` key?

File: src/inference/include/openvino/runtime/intel_gpu/params.hpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/**
+ * @brief A header for advanced hardware related properties for GPU plugin
+ *        To use in SetConfig() method of plugins
+ *
+ * @file params.hpp
```
Comment (apankratovantonp) [inline src/inference/include/openvino/runtime/intel_gpu/params.hpp:9]: may be rename to `intel_gpu/hw_properties.hpp`?

File: src/tests/functional/plugin/gpu/shared_tests_instances/behavior/ov_plugin/core_integration.cpp [hunk 4]
```diff
@@ -123,6 +124,200 @@ INSTANTIATE_TEST_SUITE_P(nightly_OVClassGetMetricTest,
...
+    std::cout << std::endl;
+
+    OV_ASSERT_PROPERTY_SUPPORTED(ov::supported_properties);
+}
+
+TEST_P(OVClassGetPropertyTest_GPU, GetMetricAvailableDevicesAndPrintNoThrow) {
+    ov::Core ie;
+    ov::Any p;
+
+    ASSERT_NO_THROW(p = ie.get_property(deviceName, ov::available_devices));
```
Comment (apankratovantonp) [inline src/tests/functional/plugin/gpu/shared_tests_instances/behavior/ov_plugin/core_integration.cpp:149]: `ie.get_property(deviceName, ov::available_devices)` call returns `std::vector<std::string>` so you don't need `ov::Any` here

Comment (apankratovantonp) [reply src/tests/functional/plugin/gpu/shared_tests_instances/behavior/ov_plugin/core_integration.cpp:149]: Please remove `ov::Any` from all other tests here if they used `ov::Property` variables

Comment (sshlyapn) [reply src/tests/functional/plugin/gpu/shared_tests_instances/behavior/ov_plugin/core_integration.cpp]: Fixed, thanks

File: src/inference/include/openvino/runtime/intel_gpu/params.hpp [hunk 5]
```diff
@@ -0,0 +1,177 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/**
+ * @brief A header for advanced hardware related properties for GPU plugin
+ *        To use in SetConfig() method of plugins
+ *
+ * @file gpu_config.hpp
```
[the remark this answers is not in the document]
Comment (sshlyapn) [reply src/inference/include/openvino/runtime/intel_gpu/params.hpp]: Or maybe `remote_properties.hpp`?

File: src/plugins/intel_gpu/src/plugin/device_config.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/plugin/device_config.cpp:119]: Right. This key was added after 2021.4 release, so we can align it with new API version (replace `GPU_MODEL_PRIORITY` with `GPU_QUEUE_PRIORITY`, `GPU_HOST_TASK_PRIORITY` and `MODEL_PRIORITY` instead)

File: src/inference/include/openvino/runtime/intel_gpu/params.hpp [hunk 6]
```diff
@@ -0,0 +1,177 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/**
+ * @brief A header for advanced hardware related properties for GPU plugin
+ *        To use in SetConfig() method of plugins
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/intel_gpu/params.hpp:7]: SetConfig => ctor of Remote objects?

File: src/inference/include/openvino/runtime/intel_gpu/params.hpp [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+namespace intel_gpu {
+
+using gpu_handle_param = void*;
+
+/**
+ * @brief Enum to define context type
+ */
+enum class ContextType {
+    OCL = 0,
+    VA_SHARED = 1,
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/intel_gpu/params.hpp:25]: documentation for enum values?

File: src/inference/include/openvino/runtime/intel_gpu/params.hpp [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+inline std::istream& operator>>(std::istream& is, ContextType& context_type) {
+    std::string str;
+    is >> str;
+    if (str == "OCL") {
+        context_type = ContextType::OCL;
+    } else if (str == "VA_SHARED") {
+        context_type = ContextType::VA_SHARED;
+    } else {
+        IE_THROW() << "Unsupported context type: " + str;
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/intel_gpu/params.hpp:48]: IE_THROW => OPENVINO_ASSERT

Comment (ilya-lavrenov) [reply src/inference/include/openvino/runtime/intel_gpu/params.hpp:48]: could you please also fix in https://github.com/openvinotoolkit/openvino/blob/master/src/inference/include/openvino/runtime/intel_gpu/ocl/ocl.hpp#L198

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+ */
+static constexpr Property<std::map<std::string, uint64_t>, PropertyMutability::RO> memory_statistics{
+    "GPU_MEMORY_STATISTICS"};
+
+/**
+ * @brief Read-only property to get maximum batch size which does not cause performance degradation due to memory swap
+ * impact.
+ */
+static constexpr Property<uint32_t, PropertyMutability::RO> max_batch_size{"GPU_MAX_BATCH_SIZE"};
+
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/intel_gpu/properties.hpp:46]: @myshevts @vladimir-paramuzov should not it be common OV property?

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+ */
+static constexpr Property<uint32_t> queue_priority{"GPU_QUEUE_PRIORITY"};
+
+/**
+ * @brief Enum to define possible host task priorities
+ */
+enum class HostTaskPriority {
+    LOW = 0,
+    HIGH = 1,
+    ANY = 2,
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/intel_gpu/properties.hpp:85]: documentation for enum values?

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+ */
+static constexpr Property<HostTaskPriority> host_task_priority{"GPU_HOST_TASK_PRIORITY"};
+}  // namespace hint
+
+namespace memory_type {
+/**
+ * @brief These keys instruct the GPU plugin to use surface/buffer memory type.
+ */
+static constexpr auto surface = "GPU_SURFACE";
+static constexpr auto buffer = "GPU_BUFFER";
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/intel_gpu/properties.hpp:130]: documentation for each vairable?

File: src/plugins/intel_gpu/src/plugin/device_config.cpp [hunk 12]
```diff
@@ -11,6 +11,7 @@
 #include "file_utils.h"
 #include "intel_gpu/plugin/device_config.hpp"
 #include "intel_gpu/plugin/itt.hpp"
+#include <openvino/runtime/intel_gpu/properties.hpp>
```
Comment (ilya-lavrenov) [inline src/plugins/intel_gpu/src/plugin/device_config.cpp:14]: `<> ` => `""`

File: src/plugins/intel_gpu/src/plugin/device_config.cpp [hunk 13]
```diff
@@ -301,7 +288,9 @@ void Config::UpdateFromMap(const std::map<std::string, std::string>& configMap)
             } else {
                 IE_THROW(NotFound) << "Unsupported KEY_CLDNN_ENABLE_FP16_FOR_QUANTIZED_MODELS flag value: " << val;
             }
-        } else if (key.compare(GPUConfigParams::KEY_GPU_MAX_NUM_THREADS) == 0) {
+        } else if (key.compare(GPUConfigParams::KEY_GPU_MAX_NUM_THREADS) == 0 ||
+                   // Should be replaced with ov::threads_num after Properties update
+                   /* key.compare(ov::inference_num_threads.name()) == 0 */ false) {
```
Comment (ilya-lavrenov) [inline src/plugins/intel_gpu/src/plugin/device_config.cpp:293]: when this update should happen? What is expected here?

File: src/inference/include/openvino/runtime/intel_gpu/params.hpp (hunk 8 above)
Comment (vladimir-paramuzov) [reply src/inference/include/openvino/runtime/intel_gpu/params.hpp:48]: `OPENVINO_ASSERT` usage would be kind of weird here, as it's supposed to check some statement, but we just need to throw an exception.
Doesn't it make sense to add macro like `OPENVINO_THROW`/`OV_THROW`?

Comment (ilya-lavrenov) [reply src/inference/include/openvino/runtime/intel_gpu/params.hpp:48]: @ilyachur could you please advice about it?
as far as I remember, we have `OPENVINO_UNREACHABLE`

File: src/inference/include/openvino/runtime/intel_gpu/params.hpp (hunk 6 above)
Comment (sshlyapn) [reply src/inference/include/openvino/runtime/intel_gpu/params.hpp]: Fixed

File: src/inference/include/openvino/runtime/intel_gpu/params.hpp (hunk 7 above)
Comment (sshlyapn) [reply src/inference/include/openvino/runtime/intel_gpu/params.hpp]: Added

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp (hunk 10 above)
Comment (sshlyapn) [reply src/inference/include/openvino/runtime/intel_gpu/properties.hpp]: Added

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp (hunk 11 above)
Comment (sshlyapn) [reply src/inference/include/openvino/runtime/intel_gpu/properties.hpp]: Added

File: src/plugins/intel_gpu/src/plugin/device_config.cpp (hunk 13 above)
Comment (sshlyapn) [reply src/plugins/intel_gpu/src/plugin/device_config.cpp]: This property was introduced in this https://github.com/openvinotoolkit/openvino/pull/9679 PR, but the PR has not been merged yet. So, I added this property myself to my PR in a new commit

File: src/plugins/intel_gpu/src/plugin/device_config.cpp (hunk 12 above)
Comment (sshlyapn) [reply src/plugins/intel_gpu/src/plugin/device_config.cpp]: Replaced

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp (hunk 9 above)
Comment (vladimir-paramuzov) [reply src/inference/include/openvino/runtime/intel_gpu/properties.hpp:46]: agree. Let's move it to the common

File: src/inference/include/openvino/runtime/properties.hpp [hunk 14]
```diff
@@ -592,6 +592,11 @@ static constexpr const int32_t NUMA = -2;
 static constexpr Property<int32_t, PropertyMutability::RW> num{"NUM_STREAMS"};
 }  // namespace streams
 
+/**
+ * @brief Maximum number of threads that can be used by executor
+ */
+static constexpr Property<int32_t, PropertyMutability::RW> num_threads{"NUM_THREADS"};
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/properties.hpp:598]: which executor? We already have `inference_num_threads` and `compilation_num_threads` - why is it not enough?

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+static constexpr Property<uint32_t> queue_priority{"GPU_QUEUE_PRIORITY"};
+
+/**
+ * @brief Enum to define possible host task priorities which is set cpu core type of TBB affinity used in load network.
+ * It is only affected on Hybrid CPUs. If the device doesn't support Hybrid CPUs, it is set to the ANY.
+ */
+enum class HostTaskPriority {
+    LOW = 0,   //!< Mapped to IStreamsExecutor::Config::LITTLE (E-cores)
+    HIGH = 1,  //!< Mapped to IStreamsExecutor::Config::BIG (P-cores)
+    ANY = 2,   //!< Mapped to IStreamsExecutor::Config::ANY (Any cores)
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/intel_gpu/properties.hpp:80]: `IStreamsExecutor::Config::ANY` is an internal API, can we have generic description w/o reference to internal API?

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp (hunk 9 above)
Comment (myshevts) [reply src/inference/include/openvino/runtime/intel_gpu/properties.hpp:46]: OPTIMAL_BATCH_SIZE is already common

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+ * as defined in https://www.khronos.org/registry/OpenCL/specs/opencl-2.1-extensions.pdf,
+ * chapter 9.19. This option should be used with an unsigned integer value (1 is lowest energy consumption)
+ * 0 means no throttle hint is set and default queue created.
+ */
+static constexpr Property<uint32_t> queue_throttle{"GPU_QUEUE_THROTTLE"};
+
+/**
+ * @brief This key instructs the GPU plugin to use the OpenCL queue priority hint
+ * as defined in https://www.khronos.org/registry/OpenCL/specs/opencl-2.1-extensions.pdf
+ * this option should be used with an unsigned integer value (1 is lowest priority)
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/intel_gpu/properties.hpp:68]: can we add how `cl_khr_throttle_hints` are mapped to these numbers?

File: src/inference/include/openvino/runtime/properties.hpp (hunk 14 above)
Comment (sshlyapn) [reply src/inference/include/openvino/runtime/properties.hpp]: Replaced with compilation_num_threads

File: docs/snippets/GPU_Metric0.cpp [hunk 17]
```diff
@@ -1,12 +1,12 @@
...
 int main() {
-using namespace InferenceEngine;
 //! [part0]
-InferenceEngine::Core core;
-auto network = core.ReadNetwork("sample.xml");
-auto exeNetwork = core.LoadNetwork(network, "GPU");
-std::map<std::string, uint64_t> statistics_map = core.GetMetric("GPU", GPU_METRIC_KEY(MEMORY_STATISTICS));
+ov::Core core;
+auto network = core.read_model("sample.xml");
+auto exeNetwork = core.compile_model(network, "GPU");
```
Comment (ilya-lavrenov) [inline docs/snippets/GPU_Metric0.cpp:8]: network => model

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp (hunk 15 above)
Comment (sshlyapn) [reply src/inference/include/openvino/runtime/intel_gpu/properties.hpp]: Replaced with `ov::hint::Priority` and updated the description of ov::intel_gpu::hint::host_task_priority property

File: docs/snippets/GPU_Metric1.cpp [hunk 18]
```diff
@@ -1,25 +1,23 @@
...
 int main() {
-using namespace InferenceEngine;
 //! [part1]
-InferenceEngine::Core core;
-CNNNetwork cnnNetwork = core.ReadNetwork("network.xml");
+ov::Core core;
+std::shared_ptr<ov::Model> network = core.read_model("sample.xml");
 uint32_t n_streams = 2;
 int64_t available_device_mem_size = 3221225472;
+std::map<std::string, ov::Any> options = {{ov::hint::model_ptr.name(), network}}; // Required. Set the address of the target network. If this is not set, the MAX_BATCH_SIZE returns 1.
```
Comment (ilya-lavrenov) [inline docs/snippets/GPU_Metric1.cpp:10]: ov::AnyMap ?

File: docs/snippets/GPU_Metric1.cpp [hunk 19]
```diff
@@ -1,25 +1,23 @@
...
+#include <openvino/runtime/core.hpp>
+#include <openvino/runtime/intel_gpu/properties.hpp>
 
 int main() {
-using namespace InferenceEngine;
 //! [part1]
-InferenceEngine::Core core;
-CNNNetwork cnnNetwork = core.ReadNetwork("network.xml");
+ov::Core core;
+std::shared_ptr<ov::Model> network = core.read_model("sample.xml");
```
Comment (ilya-lavrenov) [inline docs/snippets/GPU_Metric1.cpp:7]: network => model

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp (hunk 16 above)
Comment (sshlyapn) [reply src/inference/include/openvino/runtime/intel_gpu/properties.hpp]: Thanks, updated

File: docs/snippets/GPU_Metric1.cpp [hunk 20]
```diff
@@ -1,25 +1,23 @@
...
 uint32_t n_streams = 2;
 int64_t available_device_mem_size = 3221225472;
+std::map<std::string, ov::Any> options = {{ov::hint::model_ptr.name(), network}}; // Required. Set the address of the target network. If this is not set, the MAX_BATCH_SIZE returns 1.
+options.insert(std::make_pair(ov::streams::num.name(), n_streams)); // Optional. Set only when you want to estimate max batch size for a specific throughtput streams. Default is 1 or throughtput streams set by SetConfig.
+options.insert(std::make_pair(ov::intel_gpu::hint::available_device_mem.name(), available_device_mem_size)); // Optional. Set only when you want to limit the available device mem size.
```
Comment (ilya-lavrenov) [inline docs/snippets/GPU_Metric1.cpp:12]: why not `options[ov::intel_gpu::hint::available_device_mem] = available_device_mem_size` ?

File: docs/snippets/GPU_Metric1.cpp [hunk 21]
```diff
@@ -1,25 +1,23 @@
...
+uint32_t max_batch_size = core.get_property("GPU", ov::max_batch_size.name(), options);
 //! [part1]
 //! [part2]
-std::map<std::string, Parameter> opt = {{"MODEL_PTR", cnnNetwork.getFunction()}}; // Required. Same usage as for the MAX_BATCH_SIZE above. If not set, the OPTIONAL_BATCH_SIZE returns 1.
 // This is not entirely GPU-specific metric (so METRIC_KEY is used rather than GPU_METRIC_KEY below),
 // but the GPU is the only device that supports that at the moment.
 // For the GPU, the metric already accommodates limitation for the on-device memory that the MAX_BATCH_SIZE poses.
 // so OPTIMAL_BATCH_SIZE is always less than MAX_BATCH_SIZE. Unlike the latter it is also aligned to the power of 2.
-auto optimal_batch_size = core.GetMetric("GPU", METRIC_KEY(OPTIMAL_BATCH_SIZE), options).as<unsigned int>();
+uint32_t optimal_batch_size = core.get_property("GPU", ov::optimal_batch_size.name(), options);
```
Comment (ilya-lavrenov) [inline docs/snippets/GPU_Metric1.cpp:21]: why `ov::optimal_batch_size.name()` instead of `ov::optimal_batch_size`

File: docs/snippets/GPU_Metric1.cpp (hunk 20 above)
Comment (apankratovantonp) [reply docs/snippets/GPU_Metric1.cpp:12]: `options.insert(ov::intel_gpu::hint::available_device_mem(available_device_mem_size));` is also supported

Comment (apankratovantonp) [reply docs/snippets/GPU_Metric1.cpp:12]:
```c++
ov::AnyMap properties = {
    ov::hint::model_ptr(model), 
    ov::streams::num(n_streams), 
    ov::intel_gpu::hint::available_device_mem(available_device_mem_size)
};
```

Comment (sshlyapn) [reply docs/snippets/GPU_Metric1.cpp]: @ilya-lavrenov, in current implementation we can't use ov::Property directly as a key
@apankratovantonp, I applied your suggestion, thanks

File: docs/snippets/GPU_Metric1.cpp (hunk 21 above)
Comment (sshlyapn) [reply docs/snippets/GPU_Metric1.cpp:23]: There is no implementation for `get_property(std::string, ov::Property, ov::AnyMap)` method
For now we have only two options with optional arguments for get_property method:
`get_property(const std::string& device_name, const std::string& name, const AnyMap& arguments)`
Or:
`get_property(const std::string& deviceName, const ov::Property<T, M>& property, Args&&... args)`

File: src/inference/include/ie/gpu/gpu_config.hpp [hunk 22]
```diff
@@ -76,48 +65,6 @@ namespace GPUConfigParams {
...
- * • GPUConfigParams::GPU_QUEUE_PRIORITY_MED        - mapped to CL_QUEUE_PRIORITY_MED_KHR
- * • GPUConfigParams::GPU_QUEUE_PRIORITY_LOW        - mapped to CL_QUEUE_PRIORITY_LOW_KHR
- * • GPUConfigParams::GPI_QUEUE_PRIORITY_DEFAULT    - Not set queue priority property in cl_queue_properties
- * Host task priority
- * • GPUConfigParams::GPU_HOST_TASK_PRIORITY_HIGH   - mapped to IStreamsExecutor::Config::BIG
- * • GPUConfigParams::GPU_HOST_TASK_PRIORITY_LOW    - mapped to IStreamsExecutor::Config::LITTLE
- * • GPUConfigParams::GPU_HOST_TASK_PRIORITY_ANY    - mapped to IStreamsExecutor::Config::ANY
- */
-
-DECLARE_GPU_CONFIG_KEY(MODEL_PRIORITY);
```
Comment (ilyachur) [inline src/inference/include/ie/gpu/gpu_config.hpp:110]: Why did you remove some code from legacy API?

File: src/inference/include/openvino/runtime/core.hpp [hunk 23]
```diff
@@ -42,7 +42,7 @@ class OPENVINO_RUNTIME_API Core {
     class Impl;
     std::shared_ptr<Impl> _impl;
 
-    void get_property(const std::string& device_name, const std::string& name, ov::Any& to) const;
+    void get_property(const std::string& device_name, const std::string& name, const AnyMap& arguments, Any& to) const;
```
Comment (ilyachur) [inline src/inference/include/openvino/runtime/core.hpp:45]: Why do we need this change?

File: src/inference/include/openvino/runtime/core.hpp [hunk 24]
```diff
@@ -491,8 +491,8 @@ class OPENVINO_RUNTIME_API Core {
      * @return nothing
      */
     template <typename... Properties>
-    util::EnableIfAllProperties<void, Properties...> set_property(const std::string& device_name,
-                                                                  Properties&&... properties) {
+    util::EnableIfAllStringAny<void, Properties...> set_property(const std::string& device_name,
```
Comment (ilyachur) [inline src/inference/include/openvino/runtime/core.hpp:494]: Why do we need these changes?

File: docs/snippets/GPU_Metric1.cpp (hunk 21 above)
Comment (ilya-lavrenov) [reply docs/snippets/GPU_Metric1.cpp:23]: In the latest PR https://github.com/openvinotoolkit/openvino/pull/9993/files/943141df0d24441f452ef7662245c3ad802498bd..bee150fea9ad7bb4f9360d2547cd5bedf2254cda we have added an ability to provide property instead of it's name. Can be fixed now

File: src/inference/include/openvino/runtime/properties.hpp [hunk 25]
```diff
@@ -277,6 +275,12 @@ static constexpr Property<PerformanceMode> performance_mode{"PERFORMANCE_HINT"};
  * usually this value comes from the actual use-case (e.g. number of video-cameras, or other sources of inputs)
  */
 static constexpr Property<uint32_t> num_requests{"PERFORMANCE_HINT_NUM_REQUESTS"};
+
+/**
+ * @brief This key identifies shared pointer to the ov::Model, required for some properties (ov::max_batch_size and
+ * ov::optimal_batch_size)
+ */
+static constexpr Property<std::shared_ptr<ov::Model>> model_ptr{"MODEL_PTR"};
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/properties.hpp:283]: may be just `model`? Ptr does not add any additional value IMO

CC @vladimir-paramuzov

File: src/inference/include/ie/gpu/gpu_config.hpp (hunk 22 above)
Comment (sshlyapn) [reply src/inference/include/ie/gpu/gpu_config.hpp:110]: I I was not aware about the decision to keep not released keys in both APIs. This change is reverted, thank you

File: src/inference/include/openvino/runtime/core.hpp (hunk 23 above)
Comment (sshlyapn) [reply src/inference/include/openvino/runtime/core.hpp]: Some properties require additional arguments such as MODEL_PTR for MAX_BATCH_SIZE, so that is why we need an additional map. This has already been fixed in separate PR

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+ * - HIGH is used for CL_QUEUE_THROTTLE_HIGH_KHR OpenCL throttle hint
+ */
+using ThrottleLevel = ov::hint::Priority;
+
+/**
+ * @brief This key instructs the GPU plugin to use OpenCL queue throttle hints
+ * as defined in https://www.khronos.org/registry/OpenCL/specs/opencl-2.1-extensions.pdf,
+ * chapter 9.19. This option should be used with ov::intl_gpu::hint::ThrottleLevel values.
+ */
+static constexpr Property<ov::hint::Priority> queue_throttle{"GPU_QUEUE_THROTTLE"};
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/intel_gpu/properties.hpp:64]: minor: should be `ThrottleLevel` I suppose?
can be fixed separately if no objections in other parts of PR

**Review by ilya-lavrenov — APPROVED** [review]

**Review by vladimir-paramuzov — APPROVED** [review]
